File: Amazon_LeetCode_Python_C++/ArrayString/48_rotate_image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        tra(matrix);
        rev(matrix);
    }
    
public:
    void tra(vector<vector<int> >& matrix) {
        for (int i = 0; i < matrix.size(); i++) {
            for (int j = i + 1; j < matrix.size(); j++) {
                int tmp = matrix[j][i];
                matrix[j][i] = matrix[i][j];
                matrix[i][j] = tmp;
            }
        }
    }
    
public:
    void rev(vector<vector<int> >& matrix) {
        for (int i = 0; i < matrix.size(); i++) {
            for (int j = 0; j < matrix.size() / 2; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[i][matrix.size() - 1 - j];
                matrix[i][matrix.size() - 1 - j] = tmp;
            }
        }
    }
};

File: Amazon_LeetCode_Python_C++/ArrayString/48_rotate_image.py
class Solution:
    
    def tra(self, matrix: List[List[int]]) -> None:
        for i in range(len(matrix)):
            for j in range(i + 1, len(matrix), 1):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
                
    def ref(self, matrix: List[List[int]]) -> None:
        for i in range(len(matrix)):
            for j in range(len(matrix)//2):
                matrix[i][j], matrix[i][len(matrix) - 1 - j] = matrix[i][len(matrix) - 1 - j], matrix[i][j]
    
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.tra(matrix)
        self.ref(matrix)
